src/bun.js/bindings/JSBufferList.cpp

```diff
@@ -197,27 +223,46 @@ JSC::JSValue JSBufferList::_getBuffer(JSC::VM& vm, JSC::JSGlobalObject* lexicalG
         if (UNLIKELY(!array)) {
             return throwTypeError(lexicalGlobalObject, throwScope, "_getBuffer can only be called when all buffers are Uint8Array"_s);
         }
-        const size_t len = array->byteLength();
+        const size_t len = array->byteLength() - m_read;
         if (n < len) {
-            if (UNLIKELY(!uint8Array->setFromTypedArray(lexicalGlobalObject, offset, array, 0, n, JSC::CopyType::Unobservable))) {
+            if (UNLIKELY(!uint8Array->setFromTypedArray(lexicalGlobalObject, offset, array, m_read, n, JSC::CopyType::Unobservable))) {
                 return throwOutOfMemoryError(lexicalGlobalObject, throwScope);
             }
-            auto buffer = array->possiblySharedBuffer();
-            JSC::JSUint8Array* newArray = JSC::JSUint8Array::create(lexicalGlobalObject, subclassStructure, buffer, n, len - n);
-            iter->set(vm, this, newArray);
+            m_read += n;
             break;
         }
-        if (UNLIKELY(!uint8Array->setFromTypedArray(lexicalGlobalObject, offset, array, 0, len, JSC::CopyType::Unobservable))) {
+        if (UNLIKELY(!uint8Array->setFromTypedArray(lexicalGlobalObject, offset, array, m_read, len, JSC::CopyType::Unobservable))) {
             return throwOutOfMemoryError(lexicalGlobalObject, throwScope);
         }
+        m_read = 0;
         m_deque.removeFirst();
-        if (n == len) break;
+        if (n == len)
+            break;
         n -= len;
         offset += len;
     }
     RELEASE_AND_RETURN(throwScope, uint8Array);
 }
 
+// assumes `length() > 0` and `m_read > 0`
+inline JSC::JSValue JSBufferList::trim(JSC::VM& vm, JSC::JSGlobalObject* lexicalGlobalObject, JSC::JSValue value)
+{
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
+    JSC::JSString* str = JSC::jsDynamicCast<JSC::JSString*>(value);
+    if (str) {
+        const size_t len = str->length() - m_read;
```

Comment:
this should prob be saturating subtraction so that if it overflows, it becomes zero rather than max int32

Reply:
Doesn't `JSString::length()` return unsigned integer as well? :thinking: 

https://github.com/WebKit/WebKit/blob/1a1e7ad8f8a74fe4c00d54149795b14ee5626e3a/Source/JavaScriptCore/runtime/JSString.h#L723

Reply:
ah unsigned

same problem though, if for any case `m_read` becomes greater than `str->length()` then this would wrap

Reply:
`unsigned` is 32 bit unsigned integer, `uint32_t`

`size_t` on 64 bit is a 64 bit unsigned integer, `uint64_t`

Reply:
Between all the `n < len` it shouldn't be possible for `m_read > str->length()`, assuming of course `str->length()` stays constant.

Reply:
All four sites updated in https://github.com/oven-sh/bun/pull/1735/commits/e56d73e1d58cdb981d76d7f522bb02b49e2a33c1